Markdown for odoc without system

davesnx opened the pull request:
This PR extends odoc to support markdown-generate implemented here https://github.com/ocaml/odoc/pull/1341 and supported since odoc.3.1.0 but it currently depends on another PR https://github.com/ocaml/odoc/pull/1388 that isn't released (that's why I do the changes in nix [`7e386e6` (#12581)](https://github.com/ocaml/dune/pull/12581/commits/7e386e6de516148325d39a93a403b5a2697465d3) and If I could get another pair of eyes if it's correct)

The implementation itself tried to be very similar to the html generation, except the directory targets on a flat package.

The PR adds:

```bash
dune build @doc-markdown
```

Added the odoc version check for 3.1.0

Based on https://github.com/ocaml/dune/pull/12477

Alizter committed:
fix: avoid killing process group when not applicable

We add a way of tracking whether a job is a process group leader to the
scheduler. This allows us to avoid sending a signal to a process group
when it isn't needed.

Doing so fixes an issue with dune exec -w where rebuilds would send a
kill signal to a non-existant process group assumed to be attached to
the process we were execing.

Signed-off-by: Ali Caglayan <alizter@gmail.com>

davesnx committed:
Add doc-markdown alias

Signed-off-by: David Sancho Moreno <dsnxmoreno@gmail.com>

davesnx committed:
Add odoc version detection

Signed-off-by: David Sancho Moreno <dsnxmoreno@gmail.com>

davesnx committed:
Extend Output_format with Markdown variant

Signed-off-by: David Sancho Moreno <dsnxmoreno@gmail.com>

davesnx committed:
Implement markdown generation rules

Signed-off-by: David Sancho Moreno <dsnxmoreno@gmail.com>

davesnx committed:
Add tests for markdown documentation generation

Signed-off-by: David Sancho Moreno <dsnxmoreno@gmail.com>

Alizter commented on src/dune_rules/odoc.ml:
```diff
@@ -243,6 +259,51 @@ module Flags = struct
+module Version = struct
+  type t = int * int * int (* major * minor * patch *)
+
+  let of_string s : t option =
```
We have the `re` library available that might make this version parsing simpler.

Alizter commented on src/dune_rules/odoc.ml:
```diff
@@ -243,6 +259,51 @@ module Flags = struct
+  ;;
+
+  let compare (ma1, mi1, pa1) (ma2, mi2, pa2) =
+    match Int.compare ma1 ma2 with
```
Also since these are triples, you could use the ordering in `Tuple.T3` directly.

davesnx committed:
Use Tuple.T3.compare and re for Version

Signed-off-by: David Sancho Moreno <dsnxmoreno@gmail.com>

davesnx committed:
Use setup_scripts for odoc-markdown-version-check

Signed-off-by: David Sancho Moreno <dsnxmoreno@gmail.com>

davesnx committed:
Remove odoc and odoc-parser from dune-static-overlay

Signed-off-by: David Sancho Moreno <dsnxmoreno@gmail.com>

Alizter commented on src/dune_rules/odoc.ml:
```diff
@@ -892,6 +1128,74 @@ let setup_pkg_html_rules sctx ~pkg : unit Memo.t =
+    let target = Lib lib in
+    let* odocs = odoc_artefacts sctx target in
+    let* () =
+      (* because libraries with a package are handled in the package-level rule with the system shell script for all directory target, we skip packages *)
```
Is this comment still accurate?

Alizter commented:
Can you include a test that shows the directory structure with multiple projects (perhaps with deps) in a workspace?

Alizter commented on flake.nix:
```diff
@@ -76,10 +99,12 @@
+                };
+                utop = osuper.utop.overrideAttrs {
+                  dontGzipMan = true;
+                };
```
Was this change intentional?

davesnx committed:
Avoid adding deps when markdown isn't supported

Signed-off-by: David Sancho Moreno <dsnxmoreno@gmail.com>

davesnx committed:
Add markdown test with multiple libs

Signed-off-by: David Sancho Moreno <dsnxmoreno@gmail.com>

davesnx committed:
Remove comment on markdown rules

Signed-off-by: David Sancho Moreno <dsnxmoreno@gmail.com>

davesnx committed:
Fix comment on markdown rules

Signed-off-by: David Sancho Moreno <dsnxmoreno@gmail.com>

davesnx committed:
Format odoc rules

Signed-off-by: David Sancho Moreno <dsnxmoreno@gmail.com>

davesnx committed:
Consider Compilation_mode in markdown rules

Signed-off-by: David Sancho Moreno <dsnxmoreno@gmail.com>

davesnx committed:
Merge branch 'main' of github.com:/ocaml/dune into markdown-for-odoc-without-system

* 'main' of github.com:/ocaml/dune:
  chore: reformat some really long jq lines (#13447)
  feature: add some simple metrics (#13446)

Alizter committed:
nix: rocq flake

Signed-off-by: Ali Caglayan <alizter@gmail.com>

Alizter committed:
chore(rocq): update native tests

Signed-off-by: Ali Caglayan <alizter@gmail.com>

Alizter committed:
chore(rocq): fix broken tests

Signed-off-by: Ali Caglayan <alizter@gmail.com>

update rcoq config tests

Signed-off-by: Ali Caglayan <alizter@gmail.com>

Alizter committed:
ci: add rocq

Signed-off-by: Ali Caglayan <alizter@gmail.com>

Alizter committed:
ci: remove coq

The tests are disabled anyway

Signed-off-by: Ali Caglayan <alizter@gmail.com>

Alizter committed:
chore(nix): remove coq from flake

Signed-off-by: Ali Caglayan <alizter@gmail.com>

davesnx committed:
Merge branch 'main' into markdown-for-odoc-without-system

Signed-off-by: David Sancho <dsnxmoreno@gmail.com>

punchagan committed:
fix: Correctly set os-distribution on Windows

Currently, we return a stub value that matches the OS name on Windows.
But, the `os-distribution` needs to be `cygwin` or `msys2` when building
in Cygwin or MSYS2, respectively, to detect dependencies correctly for
various packages.

We assume that the `uname` executable is on `PATH` when running under
Cygwin or MSYS2, and use that to correctly detect the `os-distribution`.

Closes #13310

Signed-off-by: Puneeth Chaganti <punchagan@muse-amuse.in>

davesnx committed:
Remove odoc version check

Signed-off-by: David Sancho Moreno <david.sancho@ahrefs.com>

davesnx committed:
Merge branch 'main' into markdown-for-odoc-without-system

MisterDA committed:
Fix the format-dune-file action documentation

Signed-off-by: Antonin Décimo <antonin.decimo@gmail.com>

Alizter committed:
Fix the `format-dune-file` action documentation (#13540)

Follow-up to #11166.

rgrinberg committed:
feature: sandbox destruction trace event (#13538)

Signed-off-by: Rudi Grinberg <me@rgrinberg.com>

rgrinberg committed:
feature: sandbox snapshot event (#13541)

Signed-off-by: Rudi Grinberg <me@rgrinberg.com>

rgrinberg committed:
refactor: modify sandbox creation to work like the other events (#13542)

Signed-off-by: Rudi Grinberg <me@rgrinberg.com>

rgrinberg committed:
refactor: extract common sandbox tracing code (#13543)

Signed-off-by: Rudi Grinberg <me@rgrinberg.com>

rgrinberg committed:
feature: sandbox extraction event (#13544)

Signed-off-by: Rudi Grinberg <me@rgrinberg.com>

Alizter committed:
fix: Correctly set os-distribution on Windows (#13536)

Currently, we return a stub value that matches the OS name on Windows.
But, the `os-distribution` needs to be `cygwin` or `msys2` when building
in Cygwin or MSYS2, respectively, to detect dependencies correctly for
various packages.

We assume that the `uname` executable is on `PATH` when running under
Cygwin or MSYS2, and use that to correctly detect the `os-distribution`.

Closes #13310

anmonteiro committed:
test(melange): show `melange.emit` shortcoming with in-source promotion (#13537)

if you have multiple `melange.emit` stanzas in the same directory, with
different module systems but the same extension, the public libraries
will compete for the same files in `node_modules`.

Signed-off-by: Antonio Nuno Monteiro <anmonteiro@gmail.com>

rgrinberg committed:
refactor: merge [Fiber_event_bus] into [Fiber] (#13547)

Signed-off-by: Rudi Grinberg <me@rgrinberg.com>

rgrinberg committed:
refactor: remove the signature of [Dune_rpc_private] (#13548)

It causes far too much copy pasting for a signature of an internal
library

---------

Signed-off-by: Rudi Grinberg <me@rgrinberg.com>

rgrinberg committed:
refactor(csexp_rpc): split into multiple modules (#13550)

Move Socket, Session, Client, Server all into their own modules

---------

Signed-off-by: Rudi Grinberg <me@rgrinberg.com>

rgrinberg committed:
refactor(rpc): move client to own module (#13549)

Signed-off-by: Rudi Grinberg <me@rgrinberg.com>

rgrinberg committed:
refactor(rpc): move server notifications to separate module (#13552)

Signed-off-by: Rudi Grinberg <me@rgrinberg.com>

rgrinberg committed:
refactor(rpc): move dropped to trace (#13554)

This is spam in the console as clients can drop the connection whenever
they want anyway

---------

Signed-off-by: Rudi Grinberg <me@rgrinberg.com>

rgrinberg committed:
refactor(rpc): move some code outside of client functor (#13553)

Signed-off-by: Rudi Grinberg <me@rgrinberg.com>

punchagan committed:
minor: Remove program_not_found_message utility function (#13546)

The `Dune_engine.Utils` module already exposes a `program_not_found`
utility that does what would be needed in most cases

Signed-off-by: Puneeth Chaganti <punchagan@muse-amuse.in>

Alizter committed:
test(rev_store): leaking mutable state

Signed-off-by: Ali Caglayan <alizter@gmail.com>

Alizter committed:
fix: avoid killing process group when not applicable  (#12360)

We add a way of tracking whether a job is a process group leader to the
scheduler. This allows us to avoid sending a signal to a process group
when it isn't needed.

Doing so fixes an issue with dune exec -w where rebuilds would send a
kill signal to a non-existant process group assumed to be attached to
the process we were execing.

- [x] depends on #12361 
- [x] changelog
- fixes #12323

Alizter committed:
ci: reenable rocq (#13529)

In this PR we:
- add rocq to the nix flake
- fixup the output of all the rocq tests
- fixup the output of all the rocq-native tests
- re-enable the Rocq CI for both normal and native
- remove coq from the nix flake and the ci

Alizter committed:
test(rev_store): leaking mutable state (#13559)

Fetching a non-existent revision twice from the revision store will
trick it into thinking it exists and then will fail.

This is due to a half-implemented mechanism to cache unsuccessful
fetches.

I will propose one of the following in a follow up PR:
- Implement this caching correctly
- Remove this caching entirely

rgrinberg committed:
nix

Signed-off-by: Rudi Grinberg <me@rgrinberg.com>

rgrinberg committed:
refactor: remove unnecessary memo and other cleanups

Signed-off-by: Rudi Grinberg <me@rgrinberg.com>

rgrinberg committed:
chore: changes

Signed-off-by: Rudi Grinberg <me@rgrinberg.com>

rgrinberg committed:
merge changes

Signed-off-by: Rudi Grinberg <me@rgrinberg.com>

rgrinberg committed:
update

Signed-off-by: Rudi Grinberg <me@rgrinberg.com>
